enhancements/kube-apiserver/stability-api-removal-notifications.md

```diff
@@ -112,9 +134,66 @@ Value of `removed_release`:
 | 4.9               | 1.23            |                       |
 | 4.10 / 4.10 EUS   | 1.24            | TBD                   |
 
+#### DeprecatedAPIRequest Resource
+
+Patch the apiserver such that requests to a deprecated API are logged to a `DeprecatedAPIRequest` resource corresponding
+to the deprecated API. This resource can help customers determine which workloads are using deprecated APIs, including
+those which are a being removed in the next release.
+
+```go
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// DeprecatedAPIRequest tracts requests made to a deprecated API. The instance name should
+// be of the form `resource.group.version`, matching the deprecated resource.
+type DeprecatedAPIRequest struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Status DeprecatedAPIRequestStatus `json:"status,omitempty"`
+}
+
+type DeprecatedAPIRequestStatus struct {
+
+	// RemovedRelease is when the API will be removed.
+	RemovedRelease string `json:"removedRelease"`
+
+	// Sources from which users have accessed the deprecated API.
+	Sources []DeprecatedAPIRequestSource `json:"sources,omitempty"`
```

Comment:
thinking about, this should be indexed by time.  Probably a slice with 24 elements representing each hour of the day.  A little difficult to parse, but with a clear way to update entries.

```go
type DeprecatedAPIRequestStatus struct
  Conditions []metav1.Condition
  // UsageInLastHour is the usage history for the last hour. This is porcelain to make the API easier to read
  // by humans seeing if they addressed a problem.
  UsageInLastHour UsageByHour
  // UsageByHour is indexed by the hour, so 12:00AM-12:59 is in index 0, 6am-6:59am is index 6.
  UsageByHour []UsageByHour
}

type UsageByHour struct{
  // UsageByServer is keyed by the server (maybe node name?)
  UsageByServer []UsageByServer
}

type UsageByServer struct{
  ServerName string
  // LastUpdate should *always* being within the hour this is for.  This is a time indicating the last
  // moment the server is recording for, not the actual update time.
  // You will need this to properly clear data inside of an hour and then to perform proper arithmetic
  // when restarting a process part way through an hour
  LastUpdate metav1.Time
  // UsageByUser records the top ten users by number of accesses
  UsageByUser []UsageByUser
}

type UsageByUser struct{
  UserName string
  Total int
  UsageByVerb []UsageByVerb
}

type UsageByVerb struct{
  Verb string
  Total int
}
```

Reply:
```go
type UsageByServer struct{
  ServerName string
```
@deads2k It seems that all I have are source IPs, is it worth the trouble to try to reverse lookup to a server name? sources could also be clients from outside the cluster. 

```go
type UsageByServer struct{
  SourceIP string
```
Also, it seems a source might have multiple IPs if it is being routed via proxies. Events give you an array. I think I would only show the last IP (which should be the "real" one), but I can also csv all the values if it would be useful to see the use of proxies.